### PR TITLE
Implementation of a pool allocator, and a port allocator that uses it

### PR DESCRIPTION
**Pull request by justinsb**
Etcd & in-memory implementations.  Also a pool-repair strategy.

### CODE REVIEW

**Comment by justinsb** [comment]
This still isn't "wired up" to actually request allocation of the ports, but this is a fairly meaty piece of code.

I do think that the repair policy needs more work: I'm using a trick that I believe should give us a quorum read without having to change our etcd wrapper to do quorum reads (if https://github.com/coreos/etcd/issues/2837 is true).  But I'm still concerned about concurrent operations during a repair operation.  It may be that we can use the same allocation mechanism as we use for IP allocation, if we can fix the similar problems there.

cc @thockin   Probably my last PR for today; we have good coverage of the PR; the remaining pieces are mostly joining these chunks together.

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -177,6 +178,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(&s.CorsAllowedOriginList, "cors-allowed-origins", "List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching.  If this list is empty CORS will not be enabled.")
 	fs.BoolVar(&s.AllowPrivileged, "allow-privileged", s.AllowPrivileged, "If true, allow privileged containers.")
 	fs.Var(&s.PortalNet, "portal-net", "A CIDR notation IP range from which to assign portal IPs. This must not overlap with any IP ranges assigned to nodes for pods.")
+	fs.Var(&s.ServiceNodePorts, "service-node-ports", "A port range to reserve for services with NodePort visibility.")
```
Comment (thockin) [inline cmd/kube-apiserver/app/server.go]: Note that it is inclusive at both ends. maybe an example of the syntax

File: pkg/master/master.go [hunk 2]
```diff
@@ -141,6 +142,9 @@ type Config struct {
 
 	// The name of the cluster.
 	ClusterName string
+
+	// The range of ports to be assigned to services with visibility=NodePort or greater
+	NodePorts util.PortRange
```
Comment (thockin) [inline pkg/master/master.go]: ServiceNodePorts

File: pkg/master/master.go [hunk 3]
```diff
@@ -169,6 +173,7 @@ type Master struct {
 	v1beta3               bool
 	v1                    bool
 	requestContextMapper  api.RequestContextMapper
+	nodePorts             util.PortRange
```
Comment (thockin) [inline pkg/master/master.go]: serviceNodePorts

File: pkg/master/master.go [hunk 4]
```diff
@@ -194,6 +199,8 @@ type Master struct {
 	endpointRegistry  endpoint.Registry
 	portalAllocator   service.IPRegistry
 
+	nodePortAllocator *portallocator.PortAllocator
```
Comment (thockin) [inline pkg/master/master.go]: serviceNodePortAllocator - sorry, "portal" is pretty clear, but "nodePort" is not.

File: pkg/master/master.go [hunk 5]
```diff
@@ -226,6 +233,14 @@ func setDefaults(c *Config) {
 		}
 		c.PortalNet = portalNet
 	}
+	if c.NodePorts.Size == 0 {
+		// TODO: Currently no way to specify no NodePort range (do we need to allow this?)
```
Comment (thockin) [inline pkg/master/master.go]: I don't understand this TODO - didn't you define a flag up above?

File: pkg/master/master.go [hunk 6]
```diff
@@ -226,6 +233,14 @@ func setDefaults(c *Config) {
...
+		// We should probably allow this for clouds that don't require NodePort to do load-balancing (GCE)
+		// but then that breaks the strict nestedness of visibility.
+		// Review post-v1
+		c.NodePorts = util.PortRange{Base: 30000, Size: 2767}
```
Comment (thockin) [inline pkg/master/master.go]: define a variable defaultRange in this scope, to make it  a bit more obvious that these magic numbers are defaults.

File: pkg/master/master.go [hunk 7]
```diff
@@ -65,6 +65,7 @@ import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
 	ipallocator "github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/ipallocator"
 	etcdipallocator "github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/ipallocator/etcd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/portallocator"
```
Comment (thockin) [inline pkg/master/master.go]: On first read - why does this follow a different pattern than ipallocator above it?

**Comment by lavalamp** [comment]
/me Sad that we're adding another thing like ip allocator to apiserver instead of taking ip allocator out of apiserver. :(

File: pkg/registry/pool/etcd/etcd_pool_allocator.go [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	a.driver = driver
+
+	a.etcd = etcd
+	if !strings.HasSuffix(baseKey, "/") {
+		baseKey += "/"
+	}
+	a.baseKey = baseKey
+}
+
+// Try to lock the resource identified by s.
```
Comment (thockin) [inline pkg/registry/pool/etcd/etcd_pool_allocator.go]: 's' is not a symbol in this scope

**Comment by lavalamp** [comment]
What feature(s) are we missing such that if we had them, no one would feel the need to add a thing requiring mastership to apiserver?

File: pkg/registry/pool/etcd/etcd_pool_allocator.go [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+	"strings"
+)
+
+// EtcdPoolAllocator is PoolAllocator that is backed by etcd, so can be used by multiple processes
+type EtcdPoolAllocator struct {
+	driver pool.PoolDriver
+
+	lock sync.Mutex // protects 'used' and 'random'
+
+	usedCached     map[string]bool
```
Comment (thockin) [inline pkg/registry/pool/etcd/etcd_pool_allocator.go]: consider util.StringSet ?

**Comment by thockin-cc** [comment]
A sane way to signal errors back to users.  We fundamentally have 2 sorts
of allocated things today.  Things that are allocated async and fail async
(scheduling decisions, volume claim bindings, external load balancers), and
things that are allocated sync and fail sync (portal IPs, host ports).

Moving from synchronous failure (the POST transaction fails with a helpful
error) to async failure (the POST transaction succeeds and the user needs
to check back later to see if it actually succeeded) is awkward.  If we do
that, we REALLY need support for waiting for conditions or something so
users can do the equivalent of "create_service && create_consumer".

We have been a bit haphazard around this - there's not a guiding principle
that I know of.  It "makes sense" that scheduling decisions are done async
(might have to add nodes or evict jobs or ...).  Once explained, it makes
sense that volume binding is async (might have to allocate a new volume or
wait for reformatting or ...).  Allocating a portal IP seems obviously
fast, and allocating a port is even more trivial.

More than all that though, there's not really such a thing as a Service
that fails to create, today (well there is - your LB could fail, but we
already handle that badly).  Introducing Service.Status.Condition[] may be
inevitable, but it's a bumpy road and we have very little time to work out
the UX here.

**Comment by erictune** [comment]
Synchronous failure makes sense for things that probably won't ever work as POST'ed.
But, for things that could fix themselves at any moment, then async errors or pending pods makes more sense, to me at least.
Seems like port allocation is something that could fix itself at any moment. (The thing that previously consumed the port could release it).

File: pkg/registry/pool/pool_allocator.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	// TODO: We could take an owner, for safety
+	Release(key string) (bool, error)
+
+	// List all allocations, for repair.
+	ListAllocations() ([]Allocation, error)
+
+	// Checks if key is locked, for repair.
+	ReadAllocation(key string) (*Allocation, uint64, error)
+
+	// Releases allocation, if still current, for repair.
```
Comment (thockin) [inline pkg/registry/pool/pool_allocator.go]: Explain this more?  How is it different than release?

**Comment by thockin** [comment]
I accept that answer, but it doesn't make the change from "creating a
Service always succeeds" to "creating a Service always succeeds but you
need to check if it ACTUALLY succeeded" any more approachable.

File: pkg/registry/service/portallocator/port_allocator.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		glog.Errorf("PortAllocator requires at least %d ports", minPortPoolSize)
+		return nil
+	}
+
+	a := &PortAllocator{
+		portRange: *portRange,
+	}
+
+	if etcd == nil {
+		pa := &pool.MemoryPoolAllocator{}
```
Comment (thockin) [inline pkg/registry/service/portallocator/port_allocator.go]: Why do we want this, ever?

File: pkg/registry/pool/pool_allocator.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	// For tests
+	DisableRandomAllocation()
+}
+
+// Iterator over the set of items that can be members of the pool
+type PoolIterator interface {
+	Next() (string, bool)
+}
+
+type Allocation struct {
```
Comment (thockin) [inline pkg/registry/pool/pool_allocator.go]: comments

File: pkg/registry/pool/pool_allocator.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	Key     string
+	Owner   string
+	Version uint64
```
Comment (thockin) [inline pkg/registry/pool/pool_allocator.go]: explain

File: pkg/registry/pool/pool_allocator.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// Iterator over the set of items that can be members of the pool
+type PoolIterator interface {
+	Next() (string, bool)
+}
+
+type Allocation struct {
+	Key     string
+	Owner   string
```
Comment (thockin) [inline pkg/registry/pool/pool_allocator.go]: explain

File: pkg/registry/pool/memory_pool_allocator.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+// MemoryPoolAllocator is PoolAllocator that is backed by an in-memory collection,
+// can't be used if there are multiple processes allocating from the pool (cf EtcdPoolAllocator)
+type MemoryPoolAllocator struct {
+	driver         PoolDriver
+	randomAttempts int
+
+	lock   sync.Mutex // protects 'used' and 'random'
+	used   map[string]Allocation
```
Comment (thockin) [inline pkg/registry/pool/memory_pool_allocator.go]: this is really expensive - you end up storing every key twice, and you've expanded what was a bitmap into a set.  I didn't follow all of the discussion on whether the ip allocator was properly consistent, but this seems to be a pretty large departure, with not much justification (so far)

File: pkg/master/master.go (hunk 5 above)
Comment (justinsb) [inline pkg/master/master.go]: Just that the user has no way to specify an empty (zero-sized) port range: if they specify one, we overwrite it

**Comment by justinsb** [comment]
Closing in favor of #8553

**Pull request closed without merging**